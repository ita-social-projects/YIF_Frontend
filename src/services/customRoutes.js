import React from 'react';
import { Route, Redirect, useHistory } from 'react-router-dom';
import { store } from '../store/store';

export function ProtectedRoute({ user, children, allowed, ...rest }) {
  const history = useHistory();
  const { currentRole } = store.getState();

  return (
    <Route
      {...rest}
      render={({ location }) => {
        if (user && allowed.includes(currentRole.role)) return children;

        if (user && !allowed.includes(currentRole.role)) history.goBack();

        if (!user) {
          return (
            <Redirect
              to={{
                pathname: '/login',
                state: { from: location },
              }}
            />
          );
        }

        return null;
      }}
    />
  );
}

export function RedirectRoute({ user, pathname, children, ...rest }) {
  return (
    <Route
      {...rest}
      render={({ location }) => {
        if (!user) return children;

        if (user) {
          return (
            <Redirect
              to={{
                pathname,
                state: { from: location },
              }}
            />
          );
        }

        return null;
      }}
    />
  );
}
