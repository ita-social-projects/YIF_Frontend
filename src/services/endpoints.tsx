export const APIUrl: string = 'https://yifbackend.tk/api/';
//export const APIUrl: string = 'https://localhost:44324/api/';
//export const APIUrl: string = 'http://localhost:5000/api/';
//export const APIUrl: string = 'https://web-yifbackend.azurewebsites.net/api/';
