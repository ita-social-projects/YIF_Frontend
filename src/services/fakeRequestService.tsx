import {
  requestStart,
  requestSuccess,
  requestFailure,
} from '../store/reducers/requestStatus';

const fakeRequestService = () => {
  return function (dispatch: any) {
    dispatch(requestStart());
    return fetch('https://jsonplaceholder.typicode.com/todos/')
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then((json) => {
        dispatch(requestSuccess());
        console.log(json);
      })
      .catch((error) => {
        dispatch(requestFailure());
      });
  };
};
export default fakeRequestService;
