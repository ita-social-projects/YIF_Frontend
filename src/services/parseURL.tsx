const parseURL = (url: string) => {
  let urlParam: any = {};
  url
    .substring(1)
    .split('&')
    .forEach((item) => {
      let param = item.split('=');
      urlParam[param[0]] = param[1];
    });
  return urlParam;
};

export default parseURL;
