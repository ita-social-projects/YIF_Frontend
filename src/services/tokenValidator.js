import React, {
  useState,
  useEffect,
  useMemo,
  useCallback,
  useContext,
  createContext,
} from 'react';
import jwt_decode from 'jwt-decode';

const authContext = createContext({});

const isTokenExpired = (token) => {
  if (token) {
    const decoded = jwt_decode(token);
    return decoded.exp <= new Date().getTime() / 1000 ? true : false;
  }
};

function AuthProvider({ children }) {
  const [token, setToken] = useState(null);

  const isExpired = useMemo(() => isTokenExpired(token), [token]);

  useEffect(() => {
    setToken(localStorage.getItem('token'));

    //цей івент же не перезапише токен в стейті, бо компонента не перерендериться? він спрацює лише коли сторінка перезагрузиться, так і має бути?
    window.addEventListener('storage', () => {
      setToken(localStorage.getItem('token'));
    });
  }, []);

  /*   const updateToken = useCallback((token) => {
    // це має бути така проста функція?
    localStorage.setItem('token', token)    
  }, []); */

  const updateToken = useCallback(async (login, password) => {
    // чи такого плану?
    const url = 'https://yifbackend.tk/api/Authentication/LoginUser';
    let response = await fetch(url, {
      method: 'POST',
      credentials: 'include',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        login,
        password,
      }),
    });

    let result = await response.json();
    localStorage.setItem('token', result.token);
    setToken(result.token);
  }, []);

  const removeToken = useCallback(() => {
    // ця функція вішається на кнопку logout?
    // виглядає підозріло просто
    setToken(null);
    localStorage.removeItem('token');
  }, []);

  return (
    <authContext.Provider
      value={{ token, isExpired, updateToken, removeToken }}
    >
      {children}
    </authContext.Provider>
  );
}

function useAuth() {
  return useContext(authContext);
}

export { AuthProvider, useAuth };
