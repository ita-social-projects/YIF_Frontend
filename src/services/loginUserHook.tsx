import {
  loginStart,
  loginFailure,
  loginSuccess,
} from '../store/reducers/loginStatus.reducer';
import { RequestData } from '../services/requestDataFunction';

const APIUrl: string = 'https://yifbackend.tk/api/Authentication/LoginUser';

const LoginUserHook = (email: string, password: string) => {
  return function (dispatch: any) {
    const loginUser = { email, password };

    dispatch(loginStart());

    return RequestData(APIUrl, 'POST', loginUser)
      .then((json: any) => {
        if (json.statusCode === 201 || json.statusCode === 200) {
          dispatch(loginSuccess(json));
        } else {
          dispatch(loginFailure(json));
        }
      })
      .catch((error) => {
        dispatch(loginFailure(error));
      });
  };
};

export default LoginUserHook;
