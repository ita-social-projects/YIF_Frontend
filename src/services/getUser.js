import { requestSecureData } from './requestDataFunction';
import { APIUrl } from '../../src/services/endpoints';
import { store } from '../store/store';
import { setUserReducer } from '../store/reducers/setUserReducer';

export const getUser = (token) => {
  requestSecureData(`${APIUrl}Users/Current`, 'GET', token)
    .then((res) => {
      const statusCode = res.statusCode.toString();
      if (statusCode.match(/^[23]\d{2}$/)) {
        console.log('get user: good');
        store.dispatch(setUserReducer(res.data));
      } else {
        console.log(res.data.message);
        console.log('get user: bad');
      }
    })
    .catch((error) => {
      console.log(error);
      console.log('get user: bad bad');
    });
};
