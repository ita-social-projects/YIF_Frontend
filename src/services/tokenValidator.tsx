import React, {
  useState,
  useEffect,
  useCallback,
  useContext,
  createContext,
  useMemo,
} from 'react';
import jwt_decode from 'jwt-decode';
import { APIUrl } from './endpoints';
import {
  setRoleReducer,
  removeRoleReducer,
} from '../store/reducers/setRoleReducer';
import { store } from '../store/store';

type Decoded = {
  email: string;
  exp: number;
  id: string;
  roles: string;
};

type Token = string | null;

type Values = {
  token: Token;
  refreshToken: Token;
  isRefreshing: boolean;
  getToken: Function;
  updateToken: Function;
  removeToken: Function;
};

const authContext = createContext({} as Values);

function AuthProvider({ children }: any) {
  const [token, setToken] = useState<Token>(localStorage.getItem('token'));
  const [refreshToken, setRefreshToken] = useState<Token>(
    localStorage.getItem('refreshToken')
  );
  const [isRefreshing, setIsRefreshing] = useState(false);

  const updateToken = useCallback((token, refreshToken) => {
    setToken(token);
    setRefreshToken(refreshToken);
    localStorage.setItem('token', token);
    localStorage.setItem('refreshToken', refreshToken);
  }, []);

  const removeToken = useCallback(() => {
    setToken(null);
    setRefreshToken(null);
    localStorage.removeItem('token');
    localStorage.removeItem('refreshToken');
    store.dispatch(removeRoleReducer());
  }, []);

  const isTokenExpired = useCallback((token: Token) => {
    try {
      const decoded: Decoded = jwt_decode(token!);
      store.dispatch(setRoleReducer(decoded.roles[1]));
      return decoded.exp <= new Date().getTime() / 1000 ? true : false;
    } catch (error) {
      removeToken();
      return false;
    }
  }, []);

  const isExpired = useMemo(() => isTokenExpired(token), [token]);

  const getToken = useCallback(async () => {
    const url = `${APIUrl}Authentication/RefreshToken`;
    let currentToken = token;
    let currentRefreshToken = refreshToken;
    if (isTokenExpired(currentToken)) {
      try {
        setIsRefreshing(true);
        let response = await fetch(url, {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            token,
            refreshToken,
          }),
        });

        if (response.ok) {
          let result = await response.json();
          currentToken = result.token;
          currentRefreshToken = result.refreshToken;
          updateToken(currentToken, currentRefreshToken);
          setIsRefreshing(false);
        } else {
          let result = await response.json();
          removeToken();
          setIsRefreshing(false);
          console.error(result.title);
        }
      } catch (error) {
        removeToken();
        setIsRefreshing(false);
        console.error(error);
      }
    }

    return currentToken;
  }, [token, updateToken, removeToken]);

  useEffect(() => {
    if (isTokenExpired(token) && !isRefreshing) getToken();
  }, [isExpired]);

  return (
    <authContext.Provider
      value={{
        token,
        refreshToken,
        isRefreshing,
        getToken,
        updateToken,
        removeToken,
      }}
    >
      {children}
    </authContext.Provider>
  );
}

function useAuth() {
  return useContext(authContext);
}

export { AuthProvider, useAuth, authContext };
