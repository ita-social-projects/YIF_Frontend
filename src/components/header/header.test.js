import React from 'react';
import { fireEvent, render, screen, act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Header from '.';
import { AuthProvider } from '../../services/tokenValidator';

describe('header with no user', () => {
  let header;
  const links = [
    { text: 'Напрями', location: '/directions' },
    { text: 'Університети', location: '/universities' },
    { text: 'Вхід', location: '/login' },
    { text: 'Реєстрація', location: '/register' },
  ];

  beforeEach(() => {
    header = render(
      <MemoryRouter>
        <Header />
      </MemoryRouter>
    );
  });

  test.each(links)('should have %s link', (link) => {
    //ensure the text is in the dom, will throw error if can't find
    const linkDom = screen.getByText(link.text);

    expect(linkDom).toHaveAttribute('href', link.location);
  });

  test('should have logo that links to home page', () => {
    const logoDom = screen.getByTestId(/logo/i);

    //check the link location
    expect(logoDom).toHaveAttribute('href', '/');

    //check the logo text
    expect(logoDom.textContent).toBe('YIF');
  });

  test('should match snapshot', () => {
    expect(header).toMatchSnapshot();
  });
});

describe('header with logged in user', () => {
  let header;
  localStorage.setItem(
    'token',
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjliNmRkNGY2LTIzMGEtNDA4Ni05YWQ5LTQyYTZlNTEwNmJmOCIsImVtYWlsIjoicm9tYW4uYXJrLmtvQGdtYWlsLmNvbSIsInJvbGVzIjoiR3JhZHVhdGUiLCJleHAiOjE2MDkzMzA2NjR9.EqY773v1vn7_OO72pu8GKpk4ylpQ-UZn8oNQMtP7WPg'
  );

  beforeEach(() => {
    header = render(
      <AuthProvider>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </AuthProvider>
    );
  });

  test('should render properly', () => {
    expect(header).toMatchSnapshot();
  });

  test('should open dropdown content on click', () => {
    const avatar = screen.getByRole('img');

    act(() => {
      fireEvent.click(avatar);
    });
    expect(header).toMatchSnapshot();
  });

  test('should clear local storage on logout', () => {
    const avatar = screen.getByRole('img');

    act(() => {
      fireEvent.click(avatar);
    });

    const logout = screen.getByRole('button');

    act(() => {
      fireEvent.click(logout);
    });

    expect(localStorage.getItem('token')).toEqual(null);
  });
});
