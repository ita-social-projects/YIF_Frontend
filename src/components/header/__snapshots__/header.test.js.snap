// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <a
          class="logo"
          data-testid="logo"
          href="/"
        >
          YIF
        </a>
        <nav>
          <div
            class="pages"
          >
            <a
              href="/directions"
            >
              Напрями
            </a>
            <a
              href="/universities"
            >
              Університети
            </a>
          </div>
          <div
            class="entry"
          >
            <a
              class="animatedButtonTransparent"
              href="/login"
            >
              Вхід
            </a>
            <a
              class="animatedButton"
              href="/register"
            >
              Реєстрація
            </a>
          </div>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header>
      <a
        class="logo"
        data-testid="logo"
        href="/"
      >
        YIF
      </a>
      <nav>
        <div
          class="pages"
        >
          <a
            href="/directions"
          >
            Напрями
          </a>
          <a
            href="/universities"
          >
            Університети
          </a>
        </div>
        <div
          class="entry"
        >
          <a
            class="animatedButtonTransparent"
            href="/login"
          >
            Вхід
          </a>
          <a
            class="animatedButton"
            href="/register"
          >
            Реєстрація
          </a>
        </div>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
